apps/server/src/modules/provisioning/strategy/tsp/tsp.strategy.ts

```
@@ -0,0 +1,135 @@
+		const roleRefs = await this.getRoleReferencesForUser(data.externalUser);
+
+		let user: UserDO;
+		if (existingUser) {
+			// TODO EW-999: Check school change
+
+			user = await this.updateUser(existingUser, data.externalUser, roleRefs, school.id);
+		} else {
+			user = await this.createUser(data.externalUser, roleRefs, school.id);
+		}
+
+		await this.ensureAccountExists(user, data.system.systemId);
+
+		return user;
+	}
+
+	private async getRoleReferencesForUser(externalUser: ExternalUserDto): Promise<RoleReference[]> {
+		const rolesDtos = await this.roleService.findByNames(externalUser.roles || []);
+		const roleRefs = rolesDtos.map((role) => new RoleReference({ id: role.id || '', name: role.name }));
+
+		return roleRefs;
+	}
+
+	private async updateUser(
+		existingUser: UserDO,
+		externalUser: ExternalUserDto,
+		roleRefs: RoleReference[],
+		schoolId: string
+	): Promise<UserDO> {
+		existingUser.roles = roleRefs;
+		existingUser.schoolId = schoolId;
+		existingUser.firstName = externalUser.firstName || existingUser.firstName;
+		existingUser.lastName = externalUser.lastName || existingUser.lastName;
+		existingUser.email = externalUser.email || existingUser.email;
+		existingUser.birthday = externalUser.birthday;
+		const updatedUser = await this.userService.save(existingUser);
+
+		return updatedUser;
+	}
+
+	private async createUser(
+		externalUser: ExternalUserDto,
+		roleRefs: RoleReference[],
+		schoolId: string
+	): Promise<UserDO> {
+		const newUser = new UserDO({
+			roles: roleRefs,
+			schoolId,
+			firstName: externalUser.firstName || '',
+			lastName: externalUser.lastName || '',
+			email: externalUser.email || '',
+			birthday: externalUser.birthday,
+		});
+		const savedUser = await this.userService.save(newUser);
+
+		return savedUser;
+	}
+
+	private async ensureAccountExists(user: UserDO, systemId: string): Promise<void> {
+		const account = await this.accountService.findByUserId(user.id || '');
```

the userId can never be empty here. If it is empty, then it should throw an error